mava/systems/sac/ff_masac.py

```diff
@@ -0,0 +1,638 @@
+]:
+    """Create the update functions for the learner.
+
+    Args:
+        cfg: System configuration.
+        env: The environment.
+        networks: Tuple of actor and critic networks.
+        optims: Tuple of actor, critic and alpha optimisers.
+        rb: The replay buffer.
+        target_entropy: The target entropy.
+
+    Returns:
+        Tuple of (explore_fn, update_fn).
+        Explore function is used for initial exploration with random actions.
+        Update function is the main learning function, it both acts and learns.
+    """
+    actor_net, q_net = networks
+    actor_opt, q_opt, alpha_opt = optims
+
+    full_action_shape = (cfg.arch.num_envs, *env.action_spec().shape)
+    num_agents = env.num_agents
+
+    def step(
+        action: Array, obs: ObservationGlobalState, env_state: State, buffer_state: BufferState
+    ) -> Tuple[Array, State, BufferState, Dict]:
+        """Given an action, step the environment and add to the buffer."""
+        env_state, timestep = jax.vmap(env.step)(env_state, action)
+        next_obs = timestep.observation
+        rewards = timestep.reward
+        terms = ~timestep.discount.astype(bool)
+        infos = timestep.extras
+
+        real_next_obs = infos["real_next_obs"]
+
+        transition = Transition(obs, action, rewards, terms, real_next_obs)
+        buffer_state = rb.add(buffer_state, transition)
+
+        return next_obs, env_state, buffer_state, infos["episode_metrics"]
+
+    # losses:
+    def q_loss_fn(
+        q_params: QVals, obs: Array, action: Array, target: Array
+    ) -> Tuple[Array, Metrics]:
+        q1_params, q2_params = q_params
+        joint_action = get_joint_action(action)
+
+        q1_a_values = q_net.apply(q1_params, obs, joint_action)
+        q2_a_values = q_net.apply(q2_params, obs, joint_action)
+
+        q1_loss = jnp.mean(jnp.square(q1_a_values - target))
+        q2_loss = jnp.mean(jnp.square(q2_a_values - target))
+
+        loss = q1_loss + q2_loss
+        loss_info = {
+            "loss": loss,
+            "q1_loss": q1_loss,
+            "q2_loss": q2_loss,
+            "q1_a_vals": q1_a_values,
+            "q2_a_vals": q2_a_values,
+        }
+
+        return loss, loss_info
+
+    def actor_loss_fn(
+        actor_params: FrozenVariableDict,
+        obs: ObservationGlobalState,
+        actions: Array,
+        alpha: Array,
+        q_params: QVals,
+        key: chex.PRNGKey,
+    ) -> Array:
+        pi = actor_net.apply(actor_params, obs)
+        new_action = pi.sample(seed=key)
+        log_prob = pi.log_prob(new_action)
+
+        # Repeat the actions from the replay buffer such that you have (B, Ag, Ag, Ac).
+        # This gives you n_agent joint actions with the action dim kept separate.
+        # Then replace along the diagonal with the new action from the policy.
+        # This replacement means that joint_action_i will have the new action for agent_i.
```

Comment:
Can we please add a legend here for that each axis is referring to? I am not sure what Ag, and Ac are :sweat_smile: